CMakeLists.txt

```diff
@@ -0,0 +1,138 @@
+
+feature_summary(
+  FATAL_ON_MISSING_REQUIRED_PACKAGES
+  WHAT REQUIRED_PACKAGES_NOT_FOUND)
+
+# ---- Options
+
+option(OPENMP "Enable OpenMP" ON)
+option(PROFILING "Enable profiling" OFF)
+option(ASAN "Use address sanitiser" OFF)
+
+# include(CheckIPOSupported) # adds lto
+# check_ipo_supported(RESULT ipo_supported OUTPUT output)
+
+# ---- Dependencies
+
+if(OPENMP)
+  include(FindOpenMP)
+endif(OPENMP)
+
+find_package(Threads)
+set_package_properties(Threads PROPERTIES TYPE REQUIRED)
+
+# ---- Build switches
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+# set(CMAKE_INTERPROCEDURAL_OPTIMIZATION ${ipo_supported})
+
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE Release CACHE STRING
+          "Choose the type of build, options are: Release|Debug|RelWithDebInfo (for distros)." FORCE)
+endif()
+
+if (${CMAKE_BUILD_TYPE} MATCHES Release)
+  set(EXTRA_FLAGS "-march=native -D_FILE_OFFSET_BITS=64")
+endif()
+
+if ((${CMAKE_BUILD_TYPE} MATCHES Release) OR (${CMAKE_BUILD_TYPE} MATCHES RelWithDebInfo))
+  SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS} -DWFA_PARALLEL")
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS} -DWFA_PARALLEL")
+endif ()
+
+if (${CMAKE_BUILD_TYPE} MATCHES "Debug")
+  set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${EXTRA_FLAGS}")
+  set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${EXTRA_FLAGS}")
+  add_definitions(-Wfatal-errors)
+endif ()
+
+if (ASAN)
+  set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fsanitize=address  -fno-omit-frame-pointer -fno-common")
+  set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=address  -fno-omit-frame-pointer -fno-common")
+endif(ASAN)
+
+if(PROFILING)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g")
+endif(PROFILING)
+
+if(GPROF)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pg")
+endif(GPROF)
+
+# ---- Include files
+
+include_directories(wavefront)
+include_directories(utils)
+include_directories(.)
+
+file(GLOB INCLUDES
+  wavefront/*.h*
+  utils/*.h*
+  )
+
+set(wfa2lib_SOURCE
+  wavefront/wavefront_align.c
+  wavefront/wavefront_aligner.c
+  wavefront/wavefront_attributes.c
+  wavefront/wavefront_backtrace_buffer.c
+  wavefront/wavefront_backtrace.c
+  wavefront/wavefront_backtrace_offload.c
+  wavefront/wavefront_bialign.c
+  wavefront/wavefront_bialigner.c
+  wavefront/wavefront.c
+  wavefront/wavefront_components.c
+  wavefront/wavefront_compute_affine2p.c
+  wavefront/wavefront_compute_affine.c
+  wavefront/wavefront_compute.c
+  wavefront/wavefront_compute_edit.c
+  wavefront/wavefront_compute_linear.c
+  wavefront/wavefront_debug.c
+  wavefront/wavefront_display.c
+  wavefront/wavefront_extend.c
+  wavefront/wavefront_heuristic.c
+  wavefront/wavefront_pcigar.c
+  wavefront/wavefront_penalties.c
+  wavefront/wavefront_plot.c
+  wavefront/wavefront_slab.c
+  wavefront/wavefront_unialign.c
+)
+
+
+add_library(wfa-static STATIC
+    ${wfa2lib_SOURCE}
+    )
+set_target_properties(wfa-static
+  PROPERTIES OUTPUT_NAME wfa)
+
+add_library(wfa SHARED
+    ${wfa2lib_SOURCE}
+    )
```

Comment:
This doesn't allow the builder to choose whether to only build one or the other -- you have to build both, and install both, and I guess delete the one you don't want.

Reply:
Do you have an example on how it is done?

Reply:
- autotools: this is handled globally with `--enable-static`, `--enable-shared`, `--disable-static`, '--disable-shared' (by default, both `--enable-*` options are usually true)
- meson: `-Ddefault_library=shared` is the default, valid options are: `=shared`, `=static`, `=both`, targets defined as `library()` will produce whatever the build options want (there are specialized functions that force a specific mode)
- cmake: nothing is built in. Officially, cmake expects you to define `add_library(wfa ${wfa2lib_SOURCE})`, and then users who want both must configure cmake twice, once with `-DBUILD_SHARED_LIBS=ON` and once with `-DBUILD_SHARED_LIBS=OFF`, build the project twice, and install the project twice.


You can also define your own non-standard cmake options, such as `-DWFA2_SHARED_LIBS=ON -DWFA2_STATIC_LIBS=ON`, and then add if/else blocks around each library definition. You need to prefix the option names with "WFA2_" because cmake does not have options, only variables, variables can be set on the command line, and some variables are by mutual agreement intended to be used for user choices. You cannot list "options" with `--help`, and if you use subprojects then you cannot have multiple projects use the same "option" name, because they will overwrite each others' definitions or else simply fail to be able to be set per subproject. (Both autotools and meson have the concept of subproject-scoped options, autotools uses `--enable-shared=wfa2` and meson uses `-Dwfa2:default_library=shared`.)

Keep in mind when implementing `-DWFA2_SHARED_LIBS=ON -DWFA2_STATIC_LIBS=ON` that you should also check if both of them are defined as `OFF`, and emit
```cmake
message(FATAL_ERROR "at least one of shared or static libs must be enabled!")
```

Because cmake cannot know that you need to define either one or the other, and you are manually open-coding the checks because there's no builtin support -- so you *have* to handle all the edge cases yourself.